### PR TITLE
Feature flagging with autofill subfeatures

### PR DESCRIPTION
**Pull request by graeme**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/72649045549333/1203676877763836
iOS PR:  https://github.com/duckduckgo/iOS/pull/1555
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1044
What kind of version bump will this require?: Minor

**Optional**:

Tech Design URL:
https://app.asana.com/0/481882893211075/1203938569856702
https://app.asana.com/0/481882893211075/1203984214904257
CC: 

**Description**:

This PR:
- adds subfeatures support to the Privacy Config component
- moves the FeatureFlagger from iOS into BSK and updates it for the requirements gathered during this project
- moves the InternalUserDecider from iOS into BSK

To understand why I ended up at the solution I did, it's worth pointing out some requirements:
- We should be able to remotely toggle features using the privacy config
- Remote toggling shouldn't necessarily be possible while the feature is in development, only when we're close to release. Up until this point, we should stick with having them enabled internally alone by authenticating as an internal user.
- The current state of a given feature (particularly during development) may well be different in iOS vs macOS and we need a way to express that.
- Some features (autofill included) may be developed (and in the future disabled) only partially, introducing the need for nested features ("subfeatures") in the Privacy Configuration.

**Steps to test this PR**:
1. Check out the iOS & macOS PRs above
2. Check that any features configured using the privacy-config still work as expected (particularly autofill for iOS).
3. Check that any logic relying on the InternalUserDecider e.g:
    - internal updates and internal app icon on macOS
    - feature flags on iOS
4. Try changing the FeatureFlag configurations from iOS to make sure that they give the appropriate level of visibility.

**OS Testing**:

* [x] iOS 14
* [ ] iOS 15
* [x] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12
* [x] macOS 13

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Package.resolved [hunk 1]
```diff
@@ -6,8 +6,8 @@
         "repositoryURL": "https://github.com/duckduckgo/content-scope-scripts",
         "state": {
           "branch": null,
-          "revision": "ed10de3485f454ca59259266a0b3942bd21b43db",
-          "version": "3.4.1"
+          "revision": "801b7f23476f797c6eaa72b070e6c80abb82801a",
+          "version": "4.4.4"
```
Comment (samsymons) [inline Package.resolved:10]: Are there changes from C-S-S- required in this PR, or is this an accidental bump?

File: Sources/BrowserServicesKit/FeatureFlagger/FeatureFlagger.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+//        case .cookieConsent:
+//            return .internalOnly
+//        case .credentialsAutofill:
+//            return .remoteDevelopment(.subfeature(AutofillSubfeature.credentialsAutofill))
+//        case .duckPlayer:
+//            return .remoteReleasable(.feature(.duckPlayer))
+//    }
+// }
+
+
```
Comment (samsymons) [inline Sources/BrowserServicesKit/FeatureFlagger/FeatureFlagger.swift:90]: I noticed that Xcode's doc popover wasn't displaying this comment when option-clicking `FeatureFlagSourceProviding`, I think removing these spaces should fix that.

File: Tests/BrowserServicesKitTests/FeatureFlagging/DefaultFeatureFlaggerTests.swift [hunk 3]
```diff
@@ -0,0 +1,143 @@
+//
+//  FeatureFlaggerTests.swift
+//  
+//
+//  Created by Graeme Arthur on 21/03/2023.
+//
```
Comment (samsymons) [inline Tests/BrowserServicesKitTests/FeatureFlagging/DefaultFeatureFlaggerTests.swift:6]: Missing the standard file header here

File: Package.resolved (hunk 1 above)
Comment (graeme) [reply Package.resolved:10]: Accidental bump. Weird...

File: Sources/BrowserServicesKit/FeatureFlagger/FeatureFlagger.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import Foundation
+
+public protocol FeatureFlagger {
+
+/// Called from app features to determine whether a given feature is enabled.
+///
+/// `forProvider: F` takes a FeatureFlag type defined by the respective app which defines from what source it should be toggled
+/// see `FeatureFlagSourceProviding` comments below for more details
+    func isFeatureOn<F: FeatureFlagSourceProviding>(forProvider: F) -> Bool
+ }
```
Comment (samsymons) [inline Sources/BrowserServicesKit/FeatureFlagger/FeatureFlagger.swift:29]: Stray whitespace snuck in here:

```suggestion
}
```

**Review by samsymons — APPROVED** [review]
Looks good to me, nice work.

**Pull request merged**